fix: version command

limesten opened the pull request:
By submitting this pull request, I confirm that my contribution is made under the terms of the MIT license.

## Problem/Feature

- fixes issue: #124 

the command `go-blueprint version` works fine when downloading the pre-build binary from  a release.
However, if doing `go install github.com/melkeydev/go-blueprint@latest` it seems to result in an empty `GoBlueprintVersion` string.

## Description of Changes: 

lost a few hairs on my head investigating this but:
- Added a function called `getGoBlueprintVersion` 
If the version var has been set by goreleaser it will use that
else we will use `ReadBuildInfo` from the runtime/debug package (which gets the version from git tags)
also added a fallback which just prints version unknown

## Checklist

- [x] I have self-reviewed the changes being requested
- [ ] I have updated the documentation (if applicable)

Melkeydev reviewed (commented):
I am trying out this branch and I am getting the following:
```
  ~/Documents/Personal/Go/test                                                                                                                                                             
❯ go-blueprint version
Go Blueprint CLI version (devel)
```

Is this expected? I am unsure what i did

Melkeydev commented on cmd/version.go:
```diff
@@ -5,20 +5,32 @@ package cmd
 
+func getGoBlueprintVersion() string {
+	if len(GoBlueprintVersion) != 0 {
+		return GoBlueprintVersion
```
We can probably define this inline and not have a separate variable declared on line 14

limesten replied:
But GoBlueprintVersion needs to be on package level to be able to be set by the ldflag in .goreleaser.yml?
Or am I misunderstanding you?

Melkeydev commented on cmd/version.go:
```diff
@@ -5,20 +5,32 @@ package cmd
+	if info, ok := debug.ReadBuildInfo(); ok {
+		return info.Main.Version
+	}
+	return "unknown"
```
Nit: Maybe we can return something else rather than "unknown" since that is a big ambiguous

limesten replied:
true, will try to come up with something better

limesten commented:
> I am trying out this branch and I am getting the following:
> 
> ```
>   ~/Documents/Personal/Go/test                                                                                                                                                             
> ❯ go-blueprint version
> Go Blueprint CLI version (devel)
> ```
> 
> Is this expected? I am unsure what i did

For the info.Main.Version to embed the version from the git tag it needs to be installed like:
`go install github.com/melkeydev/go-blueprint@latest` (or with the specific version instead of latest)

If we install like that and then run `go version -m go-blueprint` on the binary in the bin folder we get:

<img width="1030" alt="image" src="https://github.com/Melkeydev/go-blueprint/assets/25551968/2720d5c9-a7fe-463b-bbb2-9d8ac90bb839">
<br><br>
However, if we run `go install` locally with the source code we will get `(devel)`
<br><br>

<img width="1026" alt="image" src="https://github.com/Melkeydev/go-blueprint/assets/25551968/345a38a0-7c2e-4b42-9052-78aee3c1f702">

muandane commented:
i'm a bit confused to why it's not working also at build i tested the build locally 
<img width="1022" alt="image" src="https://github.com/Melkeydev/go-blueprint/assets/114167214/1d723bef-6cf7-47f2-ad27-ea3c4b687d2e">

limesten commented:
Thoughts on these changes?

- Added more comments 

- Added an additional fallback using git revision hash and build time

- Also changed to a less ambiguous output if no version info is available

Also for additional info about getting (devel) from local builds with debug.ReadBuildInfo() there are these related issues on the golang repo:
https://github.com/golang/go/issues/29228
https://github.com/golang/go/issues/50603

Melkeydev commented:
@limesten I am still investigating this but it seems it still does not work
![image](https://github.com/Melkeydev/go-blueprint/assets/53410236/d98fca05-c41e-4f20-9595-79b24ba87cc6)

limesten commented:
> @limesten I am still investigating this but it seems it still does not work ![image](https://private-user-images.githubusercontent.com/53410236/289294393-d98fca05-c41e-4f20-9595-79b24ba87cc6.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDI1MzEyMzUsIm5iZiI6MTcwMjUzMDkzNSwicGF0aCI6Ii81MzQxMDIzNi8yODkyOTQzOTMtZDk4ZmNhMDUtYzQxZS00ZjIwLTk1OTUtNzliMjRiYTg3Y2M2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzEyMTQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjE0VDA1MTUzNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTliYmZmNDZkZDI1MWFjMjUxMDcxYzc0NjBjNDNlNTQ3NWQ2YjQ1MzcwOTRkZDlhMDliNGNkNDMyZDllMWFmZjMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.RcurwyguL1M9tPzNuWntP00sD78x1MT-7In11_1Hkkk)

I assume you are checking out the pr and then running go install. I've not found a good way to include the version info in this case. 

It will work if:

1) you make a release with goreleaser (for example goreleaser `build --single-target --skip-validate --clean` as a local test release)
(or manually enter a ldflag when running go build as @muandane did)

2) you install like this:  `go install github.com/melkeydev/go-blueprint@latest` or @v.0.4.1 etc
In this case the part after the @ is what gets added to the buildinfo so it will be accessible from the version command

The weird thing with 2) is that u cant't rly test it locally beforehand (as far as I can see at least)

Ujstor commented:
@Melkeydev This is working. I created a test repo, pushed the tag, the build passed and the version was correct. LGTM
![Screenshot from 2023-12-19 07-22-20](https://github.com/Melkeydev/go-blueprint/assets/116409846/86495998-f2f3-4baa-b2b3-a8466f32e400)

Melkeydev reviewed (approved):
LGTM
